jupyter_server/extension/manager.py

```diff
@@ -341,13 +352,30 @@ def link_extension(self, name, serverapp):
 
     def load_extension(self, name, serverapp):
         extension = self.extensions.get(name)
+
         if extension.enabled:
             try:
-                extension.load_all_points(serverapp)
-                self.log.info("{name} | extension was successfully loaded.".format(name=name))
+                points = extension.load_all_points(serverapp)
             except Exception as e:
                 self.log.debug("".join(traceback.format_exception(*sys.exc_info())))
                 self.log.warning("{name} | extension failed loading with message: {error}".format(name=name,error=str(e)))
+            else:
+                self.extension_apps.setdefault(name, set()).update((
+                    point
+                    for point in points
+                    if point is not None
+                ))
+                self.log.info("{name} | extension was successfully loaded.".format(name=name))
+
+
+    async def stop_extension(self, name, apps):
+        """Call the shutdown hooks in the specified apps."""
+        for app in apps:
+            if hasattr(app, 'stop_extension'):
+                await app.stop_extension()
+            if name in self.extension_apps:
+                # might not be the case in tests
+                del self.extension_apps[name]
```

Comment:
In general, I don’t think we need to delete stopped extensions from this map. Extensions are still there, even if they are stopped. In fact, you could imagine a scenario where someone my want to restart an extension app that has been stopped previously.

Further, this doesn’t apply if we make `self.extension_apps` a property. The value would be dynamically constructed on call.